### PR TITLE
Replace hexagon_matmul with nn_ops

### PR DESCRIPTION
**Pull request by dsharletg**
This PR removes hexagon_matmul, which was a very simple matrix multiply, and replaces it with an app nn_ops, which has several machine learning pipeline kernels.

### CODE REVIEW

File: apps/hexagon_matmul/Makefile [hunk 1]
```diff
@@ -1,23 +0,0 @@
-include ../support/Makefile.inc
```
Comment (zvookin) [inline apps/hexagon_matmul/Makefile]: If you put "all:" as the first line of the Makefile, a plain "make" will do the right thing.

**Comment by dsharletg** [comment]
Thanks to @vksnk, this now has the rest of the machine learning kernels, I updated the PR title/description.

**Comment by dsharletg** [comment]
Review ping... I just merged this with master and tested it again,

**Comment by zvookin** [comment]
This would benefit a great deal from a top-level README.md explaining what it is and probably covering goals and how to start optimizing for an existing or new platform. Mostly I'd like it to summarize that this is intended to become a central place for the fastest Halide based neural net ops across platforms Halide supports. Plus capturing any useful thoughts from working with this code and a list of things to work on.

**Comment by steven-johnson** [comment]
all of the other apps/ Makefiles have a 'make test' target; we should add one here too.

File: apps/nn_ops/AveragePool.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
+#include <assert.h>
+#include <malloc.h>
```
Comment (steven-johnson) [inline apps/nn_ops/AveragePool.cpp]: malloc.h is not present on OSX, please use stdlib.h instead

**Comment by steven-johnson** [comment]
We should really convert all of these to 'new-style' Generators, since all others are using that (aside from a testcase for legacy compatibility); among other things, new-style is necessary for effective use with the Python bindings.

(I've taken the liberty of doing the trivial conversion work in sub-branch https://github.com/halide/Halide/tree/srj-nn-ops, if you would care to take the commits :-)

**Comment by jrk** [comment]
I was going to ask “how general is this beyond Hexagon?” Skimming the code, it seems the answer is “very, given more scheduling work,” but it’s worth noting in the README that this is currently `int8` only. And related: it would be great to extend this to include float variants and backprop definitions after it lands.

**Comment by dsharletg** [comment]
Thanks @steven-johnson for the fixes, I'll work on a readme soon.

**Comment by ronlieb** [comment]
LGTM

**Pull request merged**